src/tls/openssl/tls.c

```diff
@@ -1779,3 +1825,131 @@ SSL_CTX *tls_ssl_ctx(const struct tls *tls)
 
 	return tls->ctx;
 }
+
+
+static void tls_cert_destructor(void *arg)
+{
+	struct tls_cert *uc = arg;
+
+	mem_deref(uc->host);
+	X509_free(uc->x509);
+	EVP_PKEY_free(uc->pkey);
+	sk_X509_pop_free(uc->chain, X509_free);
+}
+
+
+int tls_add_certf(struct tls *tls, const char *certf, const struct pl *host)
+{
+	struct tls_cert *uc;
+	BIO *bio = NULL;
+	int err = 0;
+
+	if (!tls || !certf)
+		return EINVAL;
+
+	uc = mem_zalloc(sizeof(*uc), tls_cert_destructor);
+	if (pl_isset(host)) {
+		err = pl_strdup(&uc->host, host);
+		if (err)
+			goto out;
+	}
+
+	bio = BIO_new_file(certf, "r");
+	if (!bio) {
+		err = EIO;
+		goto out;
+	}
+
+	uc->x509 = PEM_read_bio_X509(bio, NULL, NULL, NULL);
+	if (!uc->x509) {
+		ERR_clear_error();
+		DEBUG_WARNING("Can't read certificate from file: %s\n", certf);
+		err = ENOTSUP;
+		goto out;
+	}
+
+	while (1) {
+		X509 *ca = PEM_read_bio_X509(bio, NULL, 0, NULL);
+		if (!ca)
+			break;
+
+		if (!uc->chain)
+			uc->chain = sk_X509_new_null();
+
+		if (!uc->chain) {
+			err = ENOMEM;
+			goto out;
+		}
+
+		if (!sk_X509_push(uc->chain, ca)) {
+			err = ENOMEM;
+			goto out;
+		}
+	}
+
+	BIO_free(bio);
+	bio = BIO_new_file(certf, "r");
```

Comment:
BIO_new_file is called twice, is it needed ?

Can you use the same BIO for cert and private key ?

Reply:
I already read about `BIO_reset()`. But can't remember the result. I will test it when I have the retest.